#step2 기능구현

willing1026 opened the pull request:
#step2 입력방식 수정 ( 자동차수 -> 자동차이름 ) 
- CarNames, CarName객체 추가 ( 자동차이름 사이즈 검사 기능, 일급콜랙션 )
- CarNames, CarName을 인자로 받아서 객체 생성하도록 기존 method,  생성자 수정

#step2 - Rank객체 개선 ( List -> Map 변경. 이름별 위치 저장 )
- Rank객체 수정 : 이름별 위치를 Map에 저장하도록 변경
- Rank객체 변경에 따라 사용하는 로직 수정
- 우승자 출력 추가

willing1026 committed:
#임시커밋

willing1026 committed:
Merge branch 'sunho_step1' into sunho_step2

willing1026 committed:
#step2 입력방식 수정 ( 자동차수 -> 자동차이름 )

- CarNames, CarName객체 추가 ( 자동차이름 사이즈 검사 기능, 일급콜랙션 )
- CarNames, CarName을 인자로 받아서 객체 생성하도록 기존 method,  생성자 수정

willing1026 committed:
#step2 - Rank객체 개선 ( List -> Map 변경. 이름별 위치 저장 )

- Rank객체 수정 : 이름별 위치를 Map에 저장하도록 변경
- Rank객체 변경에 따라 사용하는 로직 수정
- 우승자 출력 추가

impactrudia commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -3,25 +3,31 @@
         this.position += distance;
     }
 
+    public String getCarName() {
```
getCarName대신에 toString()오버라이딩해서 carName반환해보는건 어떨까요? 
우린 최대한 getter쓰지 않는게 좋으니깐요~! :)

아 CarName클래스에 제가 원하던 방식을 사용하셨네요~ 그럼 이 메소드는 꼭 필요한건가요?

willing1026 replied:
결과(Rank)객체에 담을때 CarName 객체를 외부에 노출시키고 싶지 않아서 Car객체에서 getCarName() 메서드를 사용하고 리턴할때 CarName객체가 아니라 toString값을 리턴하도록 했었어요

수정했습니다!

impactrudia commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -9,6 +9,12 @@ public static Integer inputNumberOfCars() {
     }
 
+    public static String inputRacingCars() {
+        Scanner scanner = new Scanner(System.in);
```
이부분 static으로 선언해도 좋을 것 같아요. 다른 메소드에서도 중복적으로 사용하고 있어서요~

willing1026 replied:
오 감사합니다 수정했어요

impactrudia commented on src/test/java/racingcar/domain/rank/RankTest.java:
```diff
@@ -17,25 +18,20 @@
-        for (Integer position : result) {
-            assertThat(position).isEqualTo(++idx);
-        }
+        List<String> winner = rank.findWinner();
```
winnerNames가 어떨까요? 리스트 형태라서 네이밍 변경하면 좀 더 읽기 쉬울 거 같아요 :)

willing1026 replied:
수정했습니다

impactrudia commented on src/main/java/racingcar/domain/car/Car.java:
```diff
@@ -3,25 +3,31 @@
+    }
+
     public int getPosition() {
         return position;
```
강사님이 전에 이거 isMaxPosition인지 체크해서 가져오라고 했던게 기억나네요~ 
단순한 getter가 아닌 상태값을 반환해보는건 구현해보는건 어떨까요? :)

willing1026 replied:
구조가 달라서 수정불가, pass ㅠㅠ

impactrudia commented on src/main/java/racingcar/domain/car/CarName.java:
```diff
@@ -0,0 +1,22 @@
+    private String checkName(String racerName) {
+        return Optional.ofNullable(racerName)
+                .filter(name -> !name.trim().isEmpty())
+                .orElseThrow(IllegalArgumentException::new);
```
전 아직도 Optional 쓰는게 익숙하지 않아요:(
예외처리하는 부분 배워갑니다 👍

willing1026 replied:
감사합니다

impactrudia commented on src/main/java/racingcar/domain/car/CarName.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public String toString() {
+        return name;
```
Car클래스에 getCarName메서드 대신에 이걸로 사용해주세요 :)

willing1026 replied:
했나봐요

impactrudia commented on src/main/java/racingcar/domain/rank/Rank.java:
```diff
@@ -2,26 +2,47 @@
+        List<String> winner = new ArrayList<>();
+
+        result.forEach((carName, position) -> {
+            if (maxPosition == position) {
```
아까 말했던 것처럼 여기서 MaxPosition을 구했고 Car객체에에서 너가 MaxPosition가지고 있어? 라는걸로 체크해도 좋을 것 같아요! 구조 바꾸시기 힘드시면 패스하셔도 됩니다~ :)

impactrudia commented on src/test/java/racingcar/domain/car/CarNamesTest.java:
```diff
@@ -0,0 +1,20 @@
+public class CarNamesTest {
+    @Test
+    public void CarNames객체생성() {
+        //given
```
given/when/then 잘 쓰고 계시네요~ 저도 이번에 이 형식 넣어서 적용해봤어요!!
근데 줄바꿈에 대한 통일이 필요할 거 같아요.

//given
블라블라
//when
블라블라

OR
 
//given
블라블라
(줄바꿈)
//when
블라블라


줄바꿈 할지 말지 이것 중에 하나 통일해서 써보시는게 어떨까요? :)

willing1026 replied:
라인공백 넣는걸로 통일했어요!

impactrudia reviewed (commented):
Optional 함수와 Assert의 다양한 함수를 쓰신 거 배우고 갑니다.
저보다 고수이신 거 같은데 좀 지나치게 세세하다 이런 부분들은 그냥 넘어가셔도 돼요~
다음 한 주간도 화이팅!!!

willing1026 committed:
#step2 - 결과 출력방식 변경

- Ranks 객체 추가.  각 횟수별 결과를 각각 Rank객체에 저장한 뒤 Ranks 객체에 등록
- Output으로는 Ranks 객체를 넘겨서 출력
- 기타 컨벤션, 공백 등 수정

willing1026 commented:
@rudiamoon 
꼼꼼하게 피드백 잘해주셔서 감사합니다
내일한번 바꿔볼게요!  목,금 화이팅!! ㅋㅋㅋ

willing1026 committed:
#step2 피드백반영 - getter제거. toString 사용

willing1026 committed:
#step2 피드백반영 - getter제거. toString 사용

willing1026 committed:
#step2 피드백반영 - Scanner 중복사용 제거, 미사용코드 제거

impactrudia commented on src/main/java/racingcar/domain/car/Cars.java:
```diff
@@ -25,14 +26,19 @@ public static Cars newInstance(CarNames carNames, Power power) {
         return this.cars.size();
     }
+
+    public Rank move2() {
```
move2()-> move()로 합쳐봐야겠네요! :)

impactrudia commented on src/test/java/racingcar/domain/rank/RanksTest.java:
```diff
@@ -0,0 +1,55 @@
+    @Before
+    public void setUp() throws Exception {
+        //given
+        step3Cars = Arrays.asList(Car.newInstanceWithPosition(new CarName("one"), 2)
```
Rank히스토리 테스트를 이렇게 할 수있다는게 흥미롭네요!! 또 배우고 갑니다!

willing1026 replied:
감사합니다

impactrudia commented on src/main/java/racingcar/domain/rank/Ranks.java:
```diff
@@ -0,0 +1,27 @@
+        this.ranks = new ArrayList<>();
+    }
+
+
```
흐흐 할게 없을땐 Line으로도 딴지 걸기~~ 한줄로 줄여주세여~

willing1026 replied:
수정했어요ㅋㅋㅋ

impactrudia reviewed (commented):
히스토리형태로 잘해주셨네요!! 남은 부분은 작은 부분이니 같이 고쳐봐요~! 오늘도 화이팅!!

willing1026 committed:
#ste2 피드백반영 - 코드컨벤션 수정(불필요한 공백라인제거,  테스트 주석 표기), outputView 수정

willing1026 committed:
#피드백반영 - winnersName 변수명, findWinners 메서드명 변경

willing1026 committed:
#step2 피드백반영 - move2 -> move
